server/aesdsocket.c

```diff
@@ -1,242 +1,344 @@
  * Refernces: 	https://www.geeksforgeeks.org/socket-programming-cc/
  *		https://beej.us/guide/bgnet/html/#sockaddr_inman
  *		https://beej.us/guide/bgnet/html/#sockaddr_inman
  *		https://www.csd.uoc.gr/~hy556/material/tutorials/cs556-3rd-tutorial.pdf
  *		https://man7.org/linux/man-pages/man2/bind.2.html
+ *		https://man7.org/linux/man-pages/man3/strftime.3.html
+ *		https://askubuntu.com/questions/1009266/pylint3-and-pip3-broken
 */
-
 #include <stdio.h>
 #include <stdlib.h>
 #include <sys/types.h>
 #include <sys/socket.h>
 #include <errno.h>
 #include <netinet/in.h>
+#include <signal.h>
+#include <fcntl.h>
+#include <stdbool.h>
+#include <syslog.h>
+#include <net/if.h>
+#include <sys/stat.h>
+#include <string.h>
+#include <sys/queue.h>
+#include <pthread.h>
+#include <sys/time.h>
 #include <arpa/inet.h>
 #include <signal.h>
 #include <netdb.h>
 #include <unistd.h>
 #include <string.h>
 #include <netdb.h>
-#include <signal.h>
-#include <fcntl.h>
-#include <stdbool.h>
-#include <syslog.h>
-#include <net/if.h>
-
-#define MYPORT 9000
-#define BUF_SIZE 1024
-
 
 int file_fd;
 int sockfd;
-int yes=1;
+pthread_mutex_t lock;
+// structure for node
+typedef struct node
+{
+    pthread_t thread;
+    bool thr_comp_stat;
+    char ip_client[INET6_ADDRSTRLEN];
+    int clientfd;
+    
+    TAILQ_ENTRY(node) entries;
+}thread_data_t;
 
+typedef TAILQ_HEAD(head_s, node) head_t;
+head_t head;
+// stamping the time with a new line
+// RFC 2822 compliant strftime format,
+void time_stamp()
+{
+    
+    time_t clock_time;
+    // buffer to  count number of bytes
+    char time_buf[100];
+    time(&clock_time);
+    
+    // strftime - format date and time
+    strftime(time_buf, sizeof(time_buf), "timestamp: %Y-%m-%d %H:%M:%S\r\n", localtime(&clock_time));
+    
+    printf("%s", time_buf);
+    
+    lseek(file_fd, 0, SEEK_END);
+    
+    //count the number bytes written
+    int cnt_byt = strlen(time_buf);
+    int cnt_write_byt = write(file_fd, time_buf, cnt_byt);
+    
+    if (cnt_write_byt != cnt_byt)
+    {
+        printf("write of bytes Unsuccessful\n");
+        
+    }
+    
+}
+// free the memory of head and tail pointer
+void memory_free()
+{
+    thread_data_t *temp;
+    while(!TAILQ_EMPTY(&head))
+    {
+        temp = TAILQ_FIRST(&head);
+        TAILQ_REMOVE(&head, temp, entries);
+        free(temp);
+    }
+}
+// Signal Handler of the signal received
 void signal_hanlder(int signum)
 {
-    printf("Caught signal %d", signum);
-    if ((signum == SIGINT) || (signum == SIGTERM)) {
+    printf("Caught signal %d\n", signum);
+    if (signum == SIGALRM)
+    {
+        printf("SIGALRM ALERT\n");
+        time_stamp();
+        alarm(10);
+    }
+    
+    if ((signum == SIGINT) || (signum == SIGTERM)) 
+    {
+        printf("Caught signal %d\n", signum);
+        
         close(file_fd);
+        
         close(sockfd);
         remove("/var/tmp/aesdsocketdata");
-        exit(0);
+        
+        memory_free();
+        exit (0);
```

Comment:
@swapnilghonge you will leak memory with this design in some circumstances, see my other comment about this.